File: src/domain/money.ts
/**
 * نموذج المال — الوحدة الصغرى (integer minor units)
 *
 * كل مبلغ عدد صحيح بالهللة. 96.47 ر.س = 9647.
 * ممنوع منعًا باتًا استخدام الأعداد العشرية (floating point) في أي حساب مالي،
 * لأن 0.1 + 0.2 !== 0.3 في IEEE 754، وده يكسر مطابقة الرصيد بدقة الهللة.
 * التحويل للريال يحدث في طبقة العرض فقط (ARCHITECTURE.md §4).
 *
 * هذا الملف لا يستورد أي شيء. domain/ نقي تمامًا.
 */

/** مبلغ بالهللة. عدد صحيح، قد يكون سالبًا. */
export type Halalas = number

/** رمز العملة بمعيار ISO 4217. */
export type Currency = 'SAR' | 'EGP' | 'USD' | 'EUR' | 'GBP' | 'AED'

/** عدد الوحدات الصغرى في وحدة واحدة، لكل عملة. */
const MINOR_UNITS: Record<Currency, number> = {
  SAR: 100, // هللة
  EGP: 100, // قرش
  USD: 100,
  EUR: 100,
  GBP: 100,
  AED: 100,
}

export function minorUnitsOf(currency: Currency): number {
  return MINOR_UNITS[currency] ?? 100
}

/** الحد الأقصى الآمن: 2^53−1 هللة ≈ 90 تريليون ريال. أي مبلغ فوقه خطأ لا تقريب. */
const MAX_SAFE = Number.MAX_SAFE_INTEGER

export class MoneyError extends Error {
  constructor(
    message: string,
    readonly input?: string,
  ) {
    super(message)
    this.name = 'MoneyError'
  }
}

/** يتحقق أن القيمة عدد صحيح صالح كمبلغ. يرمي MoneyError عند الفشل. */
export function assertHalalas(value: unknown, context = 'مبلغ'): asserts value is Halalas {
  if (typeof value !== 'number' || !Number.isInteger(value)) {
    throw new MoneyError(`${context}: القيمة ليست عددًا صحيحًا بالهللة`, String(value))
  }
  if (!Number.isFinite(value) || Math.abs(value) > MAX_SAFE) {
    throw new MoneyError(`${context}: المبلغ خارج المدى الآمن`, String(value))
  }
}

/* ────────────────────────────── التحليل (parsing) ────────────────────────────── */

/** الأرقام العربية الهندية والفارسية إلى لاتينية، والفاصلة العشرية العربية إلى نقطة. */
export function normalizeDigits(text: string): string {
  let out = ''
  for (const ch of text) {
    const code = ch.codePointAt(0)!
    if (code >= 0x0660 && code <= 0x0669) out += String.fromCharCode(code - 0x0660 + 48)
    else if (code >= 0x06f0 && code <= 0x06f9) out += String.fromCharCode(code - 0x06f0 + 48)
    else if (ch === '٫') out += '.' // الفاصلة العشرية العربية
    else if (ch === '٬') out += ',' // فاصل الآلاف العربي
    else out += ch
  }
  return out
}

/**
 * يحوّل نصًا ماليًا إلى هللات **بدون أي عملية عشرية**.
 * يقبل: "96.47" · "1,234.5" · الأرقام العربية · "-12" · "12.4-" · "(12.40)"
 * يرفض: "" · "NaN" · "Infinity" · "1.2.3" · أكثر من خانتين كسريتين غير صفرية
 *
 * السبب في تجنب parseFloat: parseFloat("96.47") * 100 = 9646.999999999998
 */
export function parseMoney(raw: string, currency: Currency = 'SAR'): Halalas {
  if (typeof raw !== 'string') throw new MoneyError('المبلغ ليس نصًا', String(raw))

  let s = normalizeDigits(raw).trim()
  if (s === '') throw new MoneyError('المبلغ فارغ', raw)

  // فواصل الآلاف والمسافات (عادية، غير قابلة للكسر، ورفيعة)
  s = s.replace(/[\s  ,]/g, '')

  // إشارة سالبة سابقة أو لاحقة، أو أقواس محاسبية
  let negative = false
  if (/^\(.*\)$/.test(s)) {
    negative = true
    s = s.slice(1, -1)
  }
  if (s.startsWith('-') || s.startsWith('−')) {
    negative = !negative
    s = s.slice(1)
  } else if (s.startsWith('+')) {
    s = s.slice(1)
  }
  if (s.endsWith('-')) {
    negative = !negative
    s = s.slice(0, -1)
  }

  const m = /^(\d*)(?:\.(\d*))?$/.exec(s)
  if (!m || (m[1] === '' && (m[2] ?? '') === '')) {
    throw new MoneyError('صيغة المبلغ غير صالحة', raw)
  }

  const intPart = m[1] === '' ? '0' : m[1]
  const fracRaw = m[2] ?? ''

  const scale = minorUnitsOf(currency)
  const decimals = String(scale).length - 1 // 100 → 2

  // خانات كسرية زائدة: تُرفض لو غير صفرية، ولا تُقرّب بصمت
  if (fracRaw.length > decimals) {
    const extra = fracRaw.slice(decimals)
    if (/[^0]/.test(extra)) {
      throw new MoneyError(
        `المبلغ فيه أكثر من ${decimals} خانة كسرية غير صفرية — لا يُقرَّب بصمت`,
        raw,
      )
    }
  }

  const frac = fracRaw.slice(0, decimals).padEnd(decimals, '0')

  // البناء من الأرقام مباشرة — لا ضرب ولا قسمة عشرية
  const digits = intPart + frac
  if (digits.length > 16) throw new MoneyError('المبلغ خارج المدى الآمن', raw)

  const value = Number(digits)
  if (!Number.isSafeInteger(value)) throw new MoneyError('المبلغ خارج المدى الآمن', raw)

  return negative ? -value : value
}

/** مثل parseMoney لكن يرجّع null بدل ما يرمي — للاستيراد الذي يجمع الأخطاء بأرقام الصفوف. */
export function tryParseMoney(raw: string, currency: Currency = 'SAR'): Halalas | null {
  try {
    return parseMoney(raw, currency)
  } catch {
    return null
  }
}

/* ────────────────────────────── الحساب ────────────────────────────── */

export function addMoney(...amounts: Halalas[]): Halalas {
  let total = 0
  for (const a of amounts) {
    assertHalalas(a)
    total += a
  }
  assertHalalas(total, 'المجموع')
  return total
}

export function subtractMoney(a: Halalas, b: Halalas): Halalas {
  assertHalalas(a)
  assertHalalas(b)
  const r = a - b
  assertHalalas(r, 'الفرق')
  return r
}

export function negateMoney(a: Halalas): Halalas {
  assertHalalas(a)
  return -a
}

export function absMoney(a: Halalas): Halalas {
  assertHalalas(a)
  return Math.abs(a)
}

/** يجمع مبالغ قائمة. مجموع القائمة الفارغة صفر — وده صفر محسوب لا صفر مفترض. */
export function sumMoney(amounts: readonly Halalas[]): Halalas {
  return addMoney(...amounts)
}

/**
 * ضرب مبلغ في عدد صحيح (عدد الأقساط مثلاً). لا يقبل كسورًا.
 * للنسب استخدم rateOfMoney أو splitMoney.
 */
export function multiplyMoneyByInt(a: Halalas, times: number): Halalas {
  assertHalalas(a)
  if (!Number.isInteger(times)) throw new MoneyError('المضروب فيه ليس عددًا صحيحًا', String(times))
  const r = a * times
  assertHalalas(r, 'الناتج')
  return r
}

/**
 * نسبة من مبلغ (numerator/denominator)، بتقريب نصف-لأعلى-بعيدًا-عن-الصفر
 * عند حد معلن واحد. العددان صحيحان — لا يدخل عشري في الحساب.
 */
export function rateOfMoney(a: Halalas, numerator: number, denominator: number): Halalas {
  assertHalalas(a)
  if (!Number.isInteger(numerator) || !Number.isInteger(denominator)) {
    throw new MoneyError('النسبة يجب أن تكون بعددين صحيحين')
  }
  if (denominator === 0) throw new MoneyError('القسمة على صفر')
  const sign = a < 0 !== numerator < 0 ? -1 : 1
  const num = Math.abs(a) * Math.abs(numerator)
  const den = Math.abs(denominator)
  const q = Math.floor(num / den)
  const rem = num - q * den
  const rounded = rem * 2 >= den ? q + 1 : q
  const r = sign * rounded
  assertHalalas(r, 'النسبة')
  return r
}

/**
 * يقسّم مبلغًا على n حصة **بدون فقد ولا خلق هللة**.
 * المجموع دائمًا يساوي الأصل بالضبط؛ الباقي يُوزَّع هللة هللة على الحصص الأولى.
 */
export function splitMoney(a: Halalas, parts: number): Halalas[] {
  assertHalalas(a)
  if (!Number.isInteger(parts) || parts <= 0) {
    throw new MoneyError('عدد الحصص يجب أن يكون عددًا صحيحًا موجبًا', String(parts))
  }
  const sign = a < 0 ? -1 : 1
  const abs = Math.abs(a)
  const base = Math.floor(abs / parts)
  let remainder = abs - base * parts
  const out: Halalas[] = []
  for (let i = 0; i < parts; i++) {
    const extra = remainder > 0 ? 1 : 0
    if (extra) remainder--
    out.push(sign * (base + extra))
  }
  return out
}

export function compareMoney(a: Halalas, b: Halalas): -1 | 0 | 1 {
  assertHalalas(a)
  assertHalalas(b)
  return a < b ? -1 : a > b ? 1 : 0
}

export const isZeroMoney = (a: Halalas): boolean => a === 0
export const isPositiveMoney = (a: Halalas): boolean => a > 0
export const isNegativeMoney = (a: Halalas): boolean => a < 0

/**
 * مقارنة بهامش نسبي — للبحث بالمبلغ ±5% (spec/06).
 * الهامش بالألف: 50 يعني ٥٪. لا يدخل عشري.
 */
export function withinRelativeTolerance(
  value: Halalas,
  target: Halalas,
  tolerancePerThousand: number,
): boolean {
  assertHalalas(value)
  assertHalalas(target)
  const margin = rateOfMoney(absMoney(target), tolerancePerThousand, 1000)
  return absMoney(subtractMoney(value, target)) <= margin
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/Money.kt
package app.masroufy.core

import kotlin.math.abs

/**
 * نموذج المال — الوحدة الصغرى. نقل `src/domain/money.ts` بنفس السلوك ونفس رسايل الأخطاء.
 *
 * كل مبلغ عدد صحيح بالهللة (`Long`): 96.47 ر.س = 9647. **ممنوع `Double` في أي حساب مالي** (CLAUDE.md #1).
 * الحد الآمن نفس التطبيق الحالي (2^53−1) عشان المبلغ اللي بيترفض هناك يترفض هنا.
 */
typealias Halalas = Long

enum class Currency { SAR, EGP, USD, EUR, GBP, AED }

/** عدد الوحدات الصغرى في الوحدة — كلها 100 دلوقتي. */
fun minorUnitsOf(@Suppress("UNUSED_PARAMETER") currency: Currency): Long = 100

/** علامة الناقص الرياضية U+2212 — بتيجي من بعض الكشوف. */
private const val MINUS_SIGN = '−'

/** نفس `Number.MAX_SAFE_INTEGER` في التطبيق الحالي. */
const val MAX_SAFE_HALALAS: Long = 9_007_199_254_740_991L

class MoneyError(message: String, val input: String? = null) : IllegalArgumentException(message)

fun assertHalalas(value: Long, context: String = uiText(TextKey.AMOUNT_CONTEXT_DEFAULT)) {
    if (value > MAX_SAFE_HALALAS || value < -MAX_SAFE_HALALAS) {
        throw MoneyError(uiText(TextKey.MONEY_OUT_OF_RANGE_CONTEXT, context), value.toString())
    }
}

/** الأرقام العربية الهندية والفارسية للاتينية، والفاصلة العشرية العربية لنقطة، وفاصل الآلاف العربي لفاصلة. */
fun normalizeDigits(text: String): String {
    val out = StringBuilder(text.length)
    for (c in text) {
        when {
            JsText.easternDigit(c) != null -> out.append(JsText.easternDigit(c))
            c.code == 0x066B -> out.append('.') // الفاصلة العشرية العربية
            c.code == 0x066C -> out.append(',') // فاصل الآلاف العربي
            else -> out.append(c)
        }
    }
    return out.toString()
}

/**
 * نص مالي ⇒ هللات **من غير أي عملية عشرية**.
 * بيقبل: "96.47" · "1,234.5" · الأرقام العربية · "-12" · "12.4-" · "(12.40)"
 * بيرفض: "" · "NaN" · "1.2.3" · أكتر من خانتين كسريتين غير صفرية (ما بيقربش بصمت)
 */
fun parseMoney(raw: String, currency: Currency = Currency.SAR): Halalas {
    var s = JsText.trim(normalizeDigits(raw))
    if (s.isEmpty()) throw MoneyError(uiText(TextKey.MONEY_EMPTY), raw)

    // فواصل الآلاف والمسافات بأنواعها
    s = s.filterNot { JsText.isWhitespace(it) || it == ',' }

    // إشارة سالبة قبل أو بعد، أو أقواس محاسبية
    var negative = false
    if (s.length >= 2 && s.first() == '(' && s.last() == ')') {
        negative = true
        s = s.substring(1, s.length - 1)
    }
    if (s.startsWith('-') || s.startsWith(MINUS_SIGN)) {
        negative = !negative
        s = s.substring(1)
    } else if (s.startsWith('+')) {
        s = s.substring(1)
    }
    if (s.endsWith('-')) {
        negative = !negative
        s = s.dropLast(1)
    }

    val dot = s.indexOf('.')
    val intRaw = if (dot < 0) s else s.substring(0, dot)
    val fracRaw = if (dot < 0) "" else s.substring(dot + 1)
    val valid = intRaw.all(JsText::isAsciiDigit) && fracRaw.all(JsText::isAsciiDigit)
    if (!valid || (intRaw.isEmpty() && fracRaw.isEmpty())) throw MoneyError(uiText(TextKey.MONEY_BAD_FORMAT), raw)

    val intPart = intRaw.ifEmpty { "0" }
    val decimals = minorUnitsOf(currency).toString().length - 1 // 100 ⇒ 2

    if (fracRaw.length > decimals && fracRaw.substring(decimals).any { it != '0' }) {
        throw MoneyError(uiText(TextKey.MONEY_TOO_MANY_DECIMALS, decimals.toString()), raw)
    }
    val frac = fracRaw.take(decimals).padEnd(decimals, '0')

    // البنا من الأرقام مباشرة — لا ضرب ولا قسمة عشرية
    val digits = intPart + frac
    if (digits.length > 16) throw MoneyError("المبلغ خارج المدى الآمن", raw)
    val value = digits.toLong()
    if (value > MAX_SAFE_HALALAS) throw MoneyError("المبلغ خارج المدى الآمن", raw)
    return if (negative) -value else value
}

/** زي `parseMoney` بس بيرجّع null بدل ما يرمي — للاستيراد اللي بيجمع الأخطاء بأرقام الصفوف. */
fun tryParseMoney(raw: String, currency: Currency = Currency.SAR): Halalas? =
    try { parseMoney(raw, currency) } catch (_: MoneyError) { null }

fun addMoney(vararg amounts: Halalas): Halalas {
    var total = 0L
    for (a in amounts) {
        assertHalalas(a)
        total += a
    }
    assertHalalas(total, uiText(TextKey.AMOUNT_CONTEXT_TOTAL))
    return total
}

/** مجموع قايمة فاضية صفر — صفر محسوب مش مفترض. */
fun sumMoney(amounts: List<Halalas>): Halalas = addMoney(*amounts.toLongArray())

fun subtractMoney(a: Halalas, b: Halalas): Halalas {
    assertHalalas(a)
    assertHalalas(b)
    val r = a - b
    assertHalalas(r, uiText(TextKey.AMOUNT_CONTEXT_DIFFERENCE))
    return r
}

fun negateMoney(a: Halalas): Halalas { assertHalalas(a); return -a }

fun absMoney(a: Halalas): Halalas { assertHalalas(a); return abs(a) }

/** ضرب في عدد صحيح (عدد الأقساط مثلًا). للنسب: `rateOfMoney` أو `splitMoney`. */
fun multiplyMoneyByInt(a: Halalas, times: Long): Halalas {
    assertHalalas(a)
    if (times != 0L && abs(a) > MAX_SAFE_HALALAS / abs(times)) {
        throw MoneyError(uiText(TextKey.MONEY_RESULT_OUT_OF_RANGE), "$a×$times")
    }
    return a * times
}

/** نسبة (بسط/مقام) بتقريب النص لفوق بعيد عن الصفر عند حد واحد معلن. عددين صحيحين — مفيش عشري. */
fun rateOfMoney(a: Halalas, numerator: Long, denominator: Long): Halalas {
    assertHalalas(a)
    if (denominator == 0L) throw MoneyError(uiText(TextKey.MONEY_DIVIDE_BY_ZERO))
    val sign = if ((a < 0) != (numerator < 0)) -1L else 1L
    val num = abs(a) * abs(numerator)
    val den = abs(denominator)
    val q = num / den
    val rem = num - q * den
    val rounded = if (rem * 2 >= den) q + 1 else q
    val r = sign * rounded
    assertHalalas(r, uiText(TextKey.AMOUNT_CONTEXT_SHARE))
    return r
}

/** تقسيم على n حصة **من غير ما هللة تضيع أو تتخلق**؛ الباقي هللة هللة على الأوائل. */
fun splitMoney(a: Halalas, parts: Int): List<Halalas> {
    assertHalalas(a)
    if (parts <= 0) throw MoneyError(uiText(TextKey.MONEY_PARTS_POSITIVE), parts.toString())
    val sign = if (a < 0) -1L else 1L
    val total = abs(a)
    val base = total / parts
    var remainder = total - base * parts
    return List(parts) {
        val extra = if (remainder > 0) 1L else 0L
        if (extra == 1L) remainder--
        sign * (base + extra)
    }
}

fun compareMoney(a: Halalas, b: Halalas): Int {
    assertHalalas(a)
    assertHalalas(b)
    return a.compareTo(b).coerceIn(-1, 1)
}

/** مقارنة بهامش نسبي بالألف (50 = ٥٪) — للبحث بالمبلغ ±5% (spec/06). */
fun withinRelativeTolerance(value: Halalas, target: Halalas, tolerancePerThousand: Long): Boolean {
    assertHalalas(value)
    assertHalalas(target)
    val margin = rateOfMoney(absMoney(target), tolerancePerThousand, 1000)
    return absMoney(subtractMoney(value, target)) <= margin
}
